knowledge_graph_class.py

```diff
@@ -0,0 +1,94 @@
+"""Knowledge Graph"""
+from typing import Dict, List, Any
+from transformers import (
+    Text2TextGenerationPipeline,
+    AutoModelForSeq2SeqLM,
+    AutoTokenizer)
+from transformers.pipelines.text2text_generation import ReturnType
+from pydantic import Field
+
+class KnowledgeGraph():
+  """Knowledge Graph class"""
+
+  max_new_tokens: int = Field(
+    default=200, description="Maximum number of new tokens to generate."
+  )
+  predictions : List[Dict[str, Any]] = Field(default=[])
+
+  def __init__(
+    self,
+    *args,
+    model_name:str='Babelscape/rebel-large',
+    **kwargs):
+
+    self.model = EntityRelationshipExtractionPipeline(model=model_name, *args, **kwargs)
+
+  def inference(self, sentence: str) -> Dict[str, Any]:
+    """inference"""
+    self.predictions = self.model(sentence)
+
+    return self.predictions
+
+class EntityRelationshipExtractionPipeline(Text2TextGenerationPipeline):
+  """Key phrase extractor pipeline class"""
+  def __init__(self, model, *args, **kwargs):
+    super().__init__(
+      model=AutoModelForSeq2SeqLM.from_pretrained(model),
+      tokenizer=AutoTokenizer.from_pretrained(model),
+      *args,
+      **kwargs)
+
+  def _extract_entity_relationships(self, text) -> List[Dict[str, Any]]:
+    triplets = []
+    relation = ""
+    subject = ""
+    relation = ""
+    object_ = ""
+    text = text.strip()
+    current = 'x'
+    for token in text.replace("<s>", "").replace("<pad>", "").replace("</s>", "").split():
+      if token == "<triplet>":
+        current = 't'
+        if relation != '':
+          triplets.append(
+            { "subject": subject.strip(), "relation": relation.strip(), "object": object_.strip() })
+          relation = ''
+        subject = ''
+      elif token == "<subj>":
+        current = 's'
+        if relation != '':
+          triplets.append(
+            { "subject": subject.strip(), "relation": relation.strip(), "object": object_.strip() })
+        object_ = ''
+      elif token == "<obj>":
+        current = 'o'
+        relation = ''
+      else:
+        if current == 't':
+          subject += ' ' + token
+        elif current == 's':
+          object_ += ' ' + token
+        elif current == 'o':
+          relation += ' ' + token
+    if subject != '' and relation != '' and object_ != '':
+      triplets.append(
+        { "subject": subject.strip(), 'relation': relation.strip(), 'object': object_.strip() })
+    return triplets
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The `_extract_entity_relationships` method seems to have a bug. The variable `relation` is declared twice (line 43 and line 45), which might be a mistake. Also, the logic in lines 52-56 and 59-63 seems to be duplicated. Consider refactoring this code to avoid duplication and ensure correct functionality.

```diff
-  relation = ""
-  subject = ""
-  relation = ""
-  object_ = ""

+  relation = ""
+  subject = ""
+  object_ = ""

-  if token == "<triplet>":
-    current = 't'
-    if relation != '':
-      triplets.append(
-        { "subject": subject.strip(), "relation": relation.strip(), "object": object_.strip() })
-    relation = ''
-    subject = ''
-  elif token == "<subj>":
-    current = 's'
-    if relation != '':
-      triplets.append(
-        { "subject": subject.strip(), "relation": relation.strip(), "object": object_.strip() })
-    object_ = ''

+  if token in ["<triplet>", "<subj>"]:
+    current = 't' if token == "<triplet>" else 's'
+    if relation != '':
+      triplets.append(
+        { "subject": subject.strip(), "relation": relation.strip(), "object": object_.strip() })
+    relation = ''
+    subject = '' if token == "<triplet>" else subject
+    object_ = '' if token == "<subj>" else object_
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->